integration/env/new_azure_go_file_to_improve_the_diff_should_rplace_azure_go.go

```diff
@@ -0,0 +1,165 @@
+
+	EnvVarAzureGuestClientID       = "AZURE_GUEST_CLIENTID"
+	EnvVarAzureGuestClientSecret   = "AZURE_GUEST_CLIENTSECRET"
+	EnvVarAzureGuestSubscriptionID = "AZURE_GUEST_SUBSCRIPTIONID"
+	EnvVarAzureGuestTenantID       = "AZURE_GUEST_TENANTID"
+
+	EnvVarCommonDomainResourceGroup = "COMMON_DOMAIN_RESOURCE_GROUP"
+)
+
+type Azure struct {
+	CIDR             string
+	CalicoSubnetCIDR string
+	MasterSubnetCIDR string
+	VPNSubnetCIDR    string
+	WorkerSubnetCIDR string
+
+	ClientID       string
+	ClientSecret   string
+	Location       string
+	SubscriptionID string
+	TenantID       string
+
+	GuestClientID       string
+	GuestClientSecret   string
+	GuestSubscriptionID string
+	GuestTenantID       string
+
+	CommonDomainResourceGroup string
+}
+
+type azureBuilderConfig struct {
+	Logger micrologger.Logger
+
+	CircleBuildNumber uint
+}
+
+type azureBuilder struct {
+	logger micrologger.Logger
+
+	circleBuildNumber uint
+}
+
+func newAzureBuilder(config Config) (*azureBuilder, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	if config.CircleBuildNumber == 0 {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CircleBuildNumber must not be empty", config)
+	}
+
+	a := &azureBuilder{
+		logger: config.Logger,
+	}
+}
+
+func (a *azureBuilder) Build(ctx context.Context) (Azure, error) {
+	var (
+		azureCIDR             string
+		azureCalicoSubnetCIDR string
+		azureMasterSubnetCIDR string
+		azureVPNSubnetCIDR    string
+		azureWorkerSubnetCIDR string
+
+		azureClientID       string
+		azureClientSecret   string
+		azureLocation       string
+		azureSubscriptionID string
+		azureTenantID       string
+
+		azureGuestClientID       string
+		azureGuestClientSecret   string
+		azureGuestSubscriptionID string
+		azureGuestTenantID       string
+
+		commonDomainResourceGroup string
+	)
+
+	err := getEnvs(
+		getEnvOptional(EnvVarAzureCIDR, &azureCIDR),
+		getEnvOptional(EnvVarAzureCalicoSubnetCIDR, &azureCalicoSubnetCIDR),
+		getEnvOptional(EnvVarAzureMasterSubnetCIDR, &azureMasterSubnetCIDR),
+		getEnvOptional(EnvVarAzureVPNSubnetCIDR, &azureVPNSubnetCIDR),
+		getEnvOptional(EnvVarAzureWorkerSubnetCIDR, &azureWorkerSubnetCIDR),
+
+		getEnvRequired(EnvVarAzureClientID, &azureClientID),
+		getEnvRequired(EnvVarAzureClientSecret, &azureClientSecret),
+		getEnvRequired(EnvVarAzureLocation, &azureLocation),
+		getEnvRequired(EnvVarAzureSubscriptionID, &azureSubscriptionID),
+		getEnvRequired(EnvVarAzureTenantID, &azureTenantID),
+
+		getEnvRequired(EnvVarAzureGuestClientID, &azureGuestClientID),
+		getEnvRequired(EnvVarAzureGuestClientSecret, &azureGuestClientSecret),
+		getEnvRequired(EnvVarAzureGuestSubscriptionID, &azureGuestSubscriptionID),
+		getEnvRequired(EnvVarAzureGuestTenantID, &azureGuestTenantID),
+
+		getEnvRequired(EnvVarCommonDomainResourceGroup, commonDomainResourceGroup),
+	)
+	if err != nil {
+		return Azure{}, microerror.Mask(err)
+	}
```

Comment:
The alternative code which I really don't like is below. I could do stuff like 

```go
azure := Azure{}

getEnvOptional(EnvVarAzureCIDR, &azure.CIDR)
```

But there are too many cases where we will use environment variables values only in `Build` methods. 

The alternative code below. It would look much worse if optional and required envs were intertwined.

```go
        azureCIDR := getEnvOptional(EnvVarAzureCIDR)
        azureCalicoSubnetCIDR := getEnvOptional(EnvVarAzureCalicoSubnetCIDR )
        azureMasterSubnetCIDR := getEnvOptional(EnvVarAzureMasterSubnetCIDR)
        azureVPNSubnetCIDR := getEnvOptional(EnvVarAzureVPNSubnetCIDR)
        azureWorkerSubnetCIDR := getEnvVarOptional(EnvVarAzureWorkerSubnetCIDR)

        azureClientID, err := getEnvRequired(EnvVarAzureClientID)
        if err != nil {
                return Azure{}, microerror.Mask(err)
        }
        azureClientSecret, err := getEnvRequired(EnvVarAzureClientSecret)
        if err != nil {
                return Azure{}, microerror.Mask(err)
        }
        azureLocation, err := getEnvRequired(EnvVarAzureLocation)
        if err != nil {
                return Azure{}, microerror.Mask(err)
        }
        azureSubscriptionID, err := getEnvRequired(EnvVarAzureSubscriptionID)
        if err != nil {
                return Azure{}, microerror.Mask(err)
        }
        azureTenantID, err := getEnvRequired(EnvVarAzureTenantID)
        if err != nil {
                return Azure{}, microerror.Mask(err)
        }

        azureGuestClientID, err := getEnvRequired(EnvVarAzureGuestClientID)
        if err != nil {
                return Azure{}, microerror.Mask(err)
        }
        azureGuestClientSecret, err := getEnvRequired(EnvVarAzureGuestClientSecret)
        if err != nil {
                return Azure{}, microerror.Mask(err)
        }
        azureGuestSubscriptionID, err := getEnvRequired(EnvVarAzureGuestSubscriptionID)
        if err != nil {
                return Azure{}, microerror.Mask(err)
        }
        azureGuestTenantID, err := getEnvRequired(EnvVarAzureGuestTenantID)
        if err != nil {
                return Azure{}, microerror.Mask(err)
        }
```

Reply:
@xh3b4sd PTALA

Reply:
I would really like to not go with pointers. I hate it. Anything can happen with this. Instead, lets please have a clear interface. Input. Output. Easy.

Reply:
I don't follow. What can go wrong here? This is the same approach as in any command line flag library in Go.

Reply:
I am not a fan of the pattern in general. In this specific example, not so much can go wrong. I personally simply do not want to apply the reference pattern in general when I do not have to, because it forces me to think more and adds complexity in general. The value can be assigned at runtime at any given point in time and only because the function returns does not mean the value does not change anymore. This is already really painful, confusing and dangerous with the command line flag handling. I do not mind the extra error handling when this means the code is simpler and more straight forward. There will also be alternatives in further go versions. When you want to do it that way, this is fine. I am done here.

Reply:
I'm not so afraid of pointers and I encourage you to embrace them as well. I think simpler code means less errors. Even if it uses some potentially dangerous features. I've seen a code when dealing a lot of copies gives similar headaches as dealing with pointers. You need to track both almost the same way.

But I played with it and it really doesn't give much gain. So I wen with error checking in place. I just made `getEnvVarOptional` return error which is always nil just to visually align things.